screen/edit-transaction/src/main/java/com/ivy/transaction/EditTransactionViewModel.kt

```diff
@@ -773,4 +821,69 @@ class EditTransactionViewModel @Inject constructor(
             save(closeScreen)
         }
     }
+
+    private suspend fun getAllTags(): ImmutableList<Tag> =
+        tagsRepository.findAll().toImmutableList()
+
+    private fun onTagSaved(name: String) {
+        viewModelScope.launch(Dispatchers.IO) {
+            NotBlankTrimmedString.from(name.toLowerCaseLocal())
+                .onRight {
+                    val tag = with(tagMapper) { createNewTag(name = it) }
+                    tagsRepository.save(tag)
+                    this@EditTransactionViewModel.tags.value = getAllTags()
+                }
+
+            saveIfEditMode()
+        }
+    }
+
+    private fun associateTagToTransaction(selectedTag: Tag) {
+        viewModelScope.launch(Dispatchers.IO) {
+            val associatedId = AssociationId(loadedTransaction().id)
+            tagsRepository.associateTagToEntity(associatedId, selectedTag.id)
+            transactionAssociatedTags.value =
+                tagsRepository.findByAssociatedId(associatedId).toImmutableList()
+        }
+    }
+
+    private fun removeTagAssociation(selectedTag: Tag) {
+        viewModelScope.launch(Dispatchers.IO) {
+            val associatedId = AssociationId(loadedTransaction().id)
+            tagsRepository.removeTagAssociation(associatedId, selectedTag.id)
+            transactionAssociatedTags.value =
+                tagsRepository.findByAssociatedId(associatedId).toImmutableList()
+        }
+    }
+
+    private fun searchTag(query: String) {
+        viewModelScope.launch(Dispatchers.IO) {
+            tagSearchJob?.cancelAndJoin()
+            delay(tagSearchDebounceTimeInMills) // Debounce effect
+            tagSearchJob = launch(Dispatchers.IO) {
+                NotBlankTrimmedString.from(query.toLowerCaseLocal())
```

Comment:
nit: you can use `.fold()` - it's nicer and the `onRight`, `onSomething` are used for side-effects

Reply:
Yes ,
Its Indeed better Will update this in future PRs